### PR TITLE
8289186: Support predicated vector load/store operations over X86 AVX2 targets.

### PR DESCRIPTION
**Pull request by jatin-bhateja**
Hi All,

[JDK-8283667](https://bugs.openjdk.org/browse/JDK-8283667) added the support to handle masked loads on non-predicated targets by blending the loaded contents with zero vector iff unmasked portion of load does not span beyond array bounds.

X86 AVX2 offers direct predicated vector loads/store instruction for non-sub word type.

This patch adds the efficient backend implementation for predicated memory operations over int/long/float/double vectors.

Please find below the JMH micro stats with and without patch.

```
System : Intel(R) Xeon(R) Platinum 8280 CPU @ 2.70GHz [28C 2S Cascadelake Server]

Baseline:
Benchmark                                          (inSize)  (outSize)   Mode  Cnt    Score   Error   Units
LoadMaskedIOOBEBenchmark.byteLoadArrayMaskIOOBE        1026       1152  thrpt    2  712.218          ops/ms
LoadMaskedIOOBEBenchmark.doubleLoadArrayMaskIOOBE      1026       1152  thrpt    2  156.912          ops/ms
LoadMaskedIOOBEBenchmark.floatLoadArrayMaskIOOBE       1026       1152  thrpt    2  255.814          ops/ms
LoadMaskedIOOBEBenchmark.intLoadArrayMaskIOOBE         1026       1152  thrpt    2  267.688          ops/ms
LoadMaskedIOOBEBenchmark.longLoadArrayMaskIOOBE        1026       1152  thrpt    2  140.957          ops/ms
LoadMaskedIOOBEBenchmark.shortLoadArrayMaskIOOBE       1026       1152  thrpt    2  474.009          ops/ms


With Opt:
Benchmark                                          (inSize)  (outSize)   Mode  Cnt     Score   Error   Units
LoadMaskedIOOBEBenchmark.byteLoadArrayMaskIOOBE        1026       1152  thrpt    2   742.781          ops/ms
LoadMaskedIOOBEBenchmark.doubleLoadArrayMaskIOOBE      1026       1152  thrpt    2  1241.021          ops/ms
LoadMaskedIOOBEBenchmark.floatLoadArrayMaskIOOBE       1026       1152  thrpt    2  2333.311          ops/ms
LoadMaskedIOOBEBenchmark.intLoadArrayMaskIOOBE         1026       1152  thrpt    2  3258.754          ops/ms
LoadMaskedIOOBEBenchmark.longLoadArrayMaskIOOBE        1026       1152  thrpt    2  1757.192          ops/ms
LoadMaskedIOOBEBenchmark.shortLoadArrayMaskIOOBE       1026       1152  thrpt    2   472.590          ops/ms
```

Predicated memory operation over sub-word type will be handled in a subsequent patch. 

Kindly review and share your feedback.

Best Regards,
Jatin

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8289186](https://bugs.openjdk.org/browse/JDK-8289186): Support predicated vector load/store operations over X86 AVX2 targets.

### Reviewers
 * [Xiaohong Gong](https://openjdk.org/census#xgong) (@XiaohongGong - Committer) ⚠️ Review applies to [2551d741](https://git.openjdk.org/jdk/pull/9324/files/2551d74114921fbf20ecd1c9f281256dab4825f5)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9324/head:pull/9324` \
`$ git checkout pull/9324`

Update a local copy of the PR: \
`$ git checkout pull/9324` \
`$ git pull https://git.openjdk.org/jdk pull/9324/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9324`

View PR using the GUI difftool: \
`$ git pr show -t 9324`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9324.diff">https://git.openjdk.org/jdk/pull/9324.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/cpu/x86/x86.ad [hunk 1]
```diff
@@ -1753,15 +1751,23 @@ const bool Matcher::match_rule_supported_vector(int opcode, int vlen, BasicType
...
-    case Op_StoreVectorMasked:
       if (!is_LP64 || !VM_Version::supports_avx512bw()) {
         return false;
       }
       if ((size_in_bits != 512) && !VM_Version::supports_avx512vl()) {
         return false;
       }
       break;
+    case Op_LoadVectorMasked:
+      if (!VM_Version::supports_avx512bw() && (is_subword_type(bt) || UseAVX < 1)) {
```
Comment (vnkozlov) [inline src/hotspot/cpu/x86/x86.ad:1762]: With `UseAVX=0` we clear `supports_avx512bw`. So the test should be 
```
if (!VM_Version::supports_avx512bw() && is_subword_type(bt) || UseAVX < 1)
```

And may be naive question. Is VectorMaskGen is used for `mask` node  creation? If so, why to have separate support checks for `LoadVectorMasked/StoreVectorMasked`?

Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/x86.ad:1762]: Hi Vladimir, Existing expression gets benefit of short-circuiting else we will need to evaluate two expressions for truly supported case.

As of now VectorMaskGen is used for AVX3 targets, specially for partial in-lining for copy and vectorize compare and post vector loop processing. Partial inlining of copy is only enabled for sub-word types and for AVX2 we do not have sub-word handling yet,  for vectorized compare partial in-lining we use an explicit threshold i.e. array length is >= 16, thus only sub-word types and 512 bit integer species qualify this threshold.

I will be posting a subsequent patch with sub-word handling for masked load/stores over AVX2 after some performance analysis along with maskgen patterns for AVX2.

File: src/hotspot/share/opto/vectorIntrinsics.cpp [hunk 2]
```diff
@@ -305,6 +305,12 @@ bool LibraryCallKit::arch_supports_vector(int sopc, int num_elem, BasicType type
       }
     }
 
+    // We reach here only when target specific checks in match_rule_supported_vector
+    // returned a true value for these predicated opcodes.
+    if (sopc == Op_StoreVectorMasked || sopc == Op_LoadVectorMasked) {
+      return true;
+    }
+
```
[the remark this answers is not in the document]
Comment (jatin-bhateja) [reply src/hotspot/share/opto/vectorIntrinsics.cpp]: Refined the check and updated to comment.

File: src/jdk.incubator.vector/share/classes/jdk/incubator/vector/ByteVector.java [hunk 3]
```diff
@@ -3920,6 +3920,7 @@ void intoMemorySegment0Template(Class<M> maskClass, MemorySegment ms, long offse
 
     // End of low-level memory operations.
 
+    @ForceInline
```
Comment (vnkozlov) [inline src/jdk.incubator.vector/share/classes/jdk/incubator/vector/ByteVector.java:3923]: Why this change? Was it missing before or you found that based on testing?
What is criteria to add `@ForceInline`?

File: src/hotspot/share/opto/vectornode.hpp [hunk 4]
```diff
@@ -904,8 +904,7 @@ class StoreVectorMaskedNode : public StoreVectorNode {
  public:
   StoreVectorMaskedNode(Node* c, Node* mem, Node* dst, Node* src, const TypePtr* at, Node* mask)
    : StoreVectorNode(c, mem, dst, at, src) {
-    assert(mask->bottom_type()->isa_vectmask(), "sanity");
```
Comment (vnkozlov) [inline src/hotspot/share/opto/vectornode.hpp:907]: Why the assert was added before? And why you can remove it now?

File: src/hotspot/share/opto/vectorIntrinsics.cpp [hunk 5]
```diff
@@ -305,6 +305,13 @@ bool LibraryCallKit::arch_supports_vector(int sopc, int num_elem, BasicType type
       }
     }
 
+    // Since operation support check is done upfront, we reach here only when target
+    // specific checks in match_rule_supported_vector returned a true value
+    // for these masked operations over non-predicated targets.
+    if (!is_supported && (sopc == Op_StoreVectorMasked || sopc == Op_LoadVectorMasked)) {
+      return true;
+    }
```
Comment (vnkozlov) [inline src/hotspot/share/opto/vectorIntrinsics.cpp:313]: Still unclear for me. As I understand `the upfront checks`, you mention, are checks at lines 270 and 286. So we know that `StoreVectorMasked` and `LoadVectorMasked` are supported.
The second part of your comment is talking about `non-predicated targets`. Does it mean that the real check should be next?:
```
   if (Matcher::has_predicated_vectors()) {
      ....
   } else if (sopc == Op_StoreVectorMasked || sopc == Op_LoadVectorMasked) {
     // your comment
     return true;
   }
   ```
   Or I am still missing what are you trying to do here?

File: src/hotspot/cpu/x86/x86.ad [hunk 6]
```diff
@@ -1753,15 +1751,23 @@ const bool Matcher::match_rule_supported_vector(int opcode, int vlen, BasicType
...
+        return false;
+      }
+      break;
+    case Op_StoreVectorMasked:
+      if (!VM_Version::supports_avx512bw() && (is_subword_type(bt) || UseAVX < 1)) {
```
Comment (vnkozlov) [inline src/hotspot/cpu/x86/x86.ad:1767]: Why duplicate the same check?
And you did not answer my suggestion about modifying the check.

File: test/micro/org/openjdk/bench/jdk/incubator/vector/LoadMaskedIOOBEBenchmark.java [hunk 7]
```diff
@@ -95,47 +95,47 @@ public void Setup() {
     public void byteLoadArrayMaskIOOBE() {
         for (int i = 0; i < inSize; i += bspecies.length()) {
             VectorMask<Byte> mask = VectorMask.fromArray(bspecies, m, i);
-            ByteVector.fromArray(bspecies, byteIn, i, mask).intoArray(byteOut, i);
+            ByteVector.fromArray(bspecies, byteIn, i, mask).intoArray(byteOut, i, mask);
```
Comment (XiaohongGong) [inline test/micro/org/openjdk/bench/jdk/incubator/vector/LoadMaskedIOOBEBenchmark.java:98]: Could you please add new benchmarks for masked `store` ?

File: src/hotspot/share/opto/vectorIntrinsics.cpp (hunk 5 above)
Comment (XiaohongGong) [reply src/hotspot/share/opto/vectorIntrinsics.cpp:313]: Could we change the `VectorMaskUseType` here https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/vectorIntrinsics.cpp#L1221 instead of doing the modification here? Originally we assume `LoadVectorMasked/StoreVectorMasked` is implemented with predicated feature, we added the `VectorMaskUsePred`. If the assumption is not changed now, we can just remote `VectorMaskUsePred` and so only `match_rule_supported_vector` is checked for such ops.

File: src/hotspot/share/opto/vectornode.hpp (hunk 4 above)
Comment (XiaohongGong) [reply src/hotspot/share/opto/vectornode.hpp:907]: Does this mean the `mask` input can be a normal `vect_type ` like the mask input of `VectorBlend` for `LoadVectorMasked/StoreVectorMasked` over X86 AVX2 systems? They do not depend on the predicated feature for the AVX2 systems, right?

File: src/hotspot/share/opto/vectorIntrinsics.cpp (hunk 5 above)
Comment (jatin-bhateja) [reply src/hotspot/share/opto/vectorIntrinsics.cpp:313]: > 

Hi Vladimir,
L220 is checking for backend support for these IR using match_rule_supported_vector, if targets do not support them then we exit early,  LoadVectorMasked and StoreVectorMasked are special IR node which were added only for predicated targets.

I agree with @XiaohongGong that we can remove this constraint for masked memory operations from the caller side i.e. LibraryCallKit::inline_vector_mem_masked_operation,  such a constraint is only useful for shared IR nodes which can carry additional mask edge on common IR for predicated targets, since there we need to check for existence of blend + vector op as a fall back case.

File: src/hotspot/share/opto/vectornode.hpp (hunk 4 above)
Comment (jatin-bhateja) [reply src/hotspot/share/opto/vectornode.hpp:907]: IR was specifically added for predicated targets (AVX512 and ARM's SVE),  this patch is re-using this IR for non-predicated AVX2 target where mask could be a vector type.

File: test/micro/org/openjdk/bench/jdk/incubator/vector/LoadMaskedIOOBEBenchmark.java (hunk 7 above)
Comment (jatin-bhateja) [reply test/micro/org/openjdk/bench/jdk/incubator/vector/LoadMaskedIOOBEBenchmark.java]: Done.

Comment (jatin-bhateja) [reply test/micro/org/openjdk/bench/jdk/incubator/vector/LoadMaskedIOOBEBenchmark.java]: Here are results of new benchmark.

```
BaseLine:
Benchmark                                            (inSize)  (outSize)   Mode  Cnt    Score   Error   Units
StoreMaskedIOOBEBenchmark.byteStoreArrayMaskIOOBE        1024       1022  thrpt    2  772.555          ops/ms
StoreMaskedIOOBEBenchmark.doubleStoreArrayMaskIOOBE      1024       1022  thrpt    2  180.548          ops/ms
StoreMaskedIOOBEBenchmark.floatStoreArrayMaskIOOBE       1024       1022  thrpt    2  311.500          ops/ms
StoreMaskedIOOBEBenchmark.intStoreArrayMaskIOOBE         1024       1022  thrpt    2  312.457          ops/ms
StoreMaskedIOOBEBenchmark.longStoreArrayMaskIOOBE        1024       1022  thrpt    2  181.013          ops/ms
StoreMaskedIOOBEBenchmark.shortStoreArrayMaskIOOBE       1024       1022  thrpt    2  538.537          ops/ms


WithOpt:

Benchmark                                            (inSize)  (outSize)   Mode  Cnt     Score   Error   Units
StoreMaskedIOOBEBenchmark.byteStoreArrayMaskIOOBE        1024       1022  thrpt    2   757.079          ops/ms
StoreMaskedIOOBEBenchmark.doubleStoreArrayMaskIOOBE      1024       1022  thrpt    2  1553.923          ops/ms
StoreMaskedIOOBEBenchmark.floatStoreArrayMaskIOOBE       1024       1022  thrpt    2  3060.020          ops/ms
StoreMaskedIOOBEBenchmark.intStoreArrayMaskIOOBE         1024       1022  thrpt    2  3025.225          ops/ms
StoreMaskedIOOBEBenchmark.longStoreArrayMaskIOOBE        1024       1022  thrpt    2  1562.263          ops/ms
StoreMaskedIOOBEBenchmark.shortStoreArrayMaskIOOBE       1024       1022  thrpt    2   538.931          ops/ms
```

**Review by XiaohongGong — APPROVED** [review]
Common IR changes look good to me!

File: src/jdk.incubator.vector/share/classes/jdk/incubator/vector/ByteVector.java (hunk 3 above)
Comment (vnkozlov) [reply src/jdk.incubator.vector/share/classes/jdk/incubator/vector/ByteVector.java]: okay.

File: src/hotspot/share/opto/vectornode.hpp (hunk 4 above)
Comment (vnkozlov) [reply src/hotspot/share/opto/vectornode.hpp:907]: Got it.

**Review by vnkozlov — COMMENTED** [review]
Good. Lat me test it.

**Comment by vnkozlov** [comment]
@jatin-bhateja, please update it to latest sources. I have conflict while applying patch.

**Comment by vnkozlov** [comment]
I started testing.

**Review by vnkozlov — APPROVED** [review]
Testing results are good

**Comment by jatin-bhateja** [comment]
Thanks @vnkozlov , @XiaohongGong for reviews.
